### PR TITLE
IPA over Grumpkin

### PR DESCRIPTION
**Pull request by storojs72**
This PR implements InnerProductArgument PCS over Grumpkin. 

The implementation is quite dense, so please review it carefully.

The reference one is: https://github.com/lurk-lab/arecibo/tree/d65918097724e7b9e3d63e7e7db55c43d9e6e7cd 

TBD in next PR: https://github.com/lurk-lab/solidity-verifier/issues/52

### CODE REVIEW

File: src/blocks/IpaPcs.sol [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+        label = new uint8[](1);
+        label[0] = 0x72;
+        uint256 r;
+        (transcript, r) = KeccakTranscriptLib.squeeze(transcript, ScalarFromUniformLib.curveGrumpkin(), label);
+
+        uint256[] memory r_vec = new uint256[](input.L_vec.length);
+        for (uint256 index = 0; index < r_vec.length; index++) {
+            // b"L" in Rust
+            label[0] = 0x4c;
+            transcript = KeccakTranscriptLib.absorb(transcript, label, input.L_vec[index].x);
```
Comment (tchataigner) [inline src/blocks/IpaPcs.sol:294]: So in `arecibo`, I could find the implementation of `Keccak256Transcript::absorb` and the use of  `to_transcript_bytes` but I could not find the implementation for the latter for `CompressedCommitment`.

Why is it that we only need to absorb the `x` of the curve point and not `x`+`y`+`infinity_byte` as we are in a lot of our implementations in arecibo?

**Review by tchataigner — COMMENTED** [review]
That looks good to me! I just have a question on the `L` and `R` steps of the verification.

File: src/blocks/IpaPcs.sol (hunk 1 above)
Comment (tchataigner) [reply src/blocks/IpaPcs.sol:294]: > The implementation of to_transcript_bytes for CompressedCommitment is indeed hidden in [following](https://github.com/lurk-lab/arecibo/blob/dev/src/provider/traits.rs#L161) macro.

Thanks! I understand now that it trace all the way to the defined `Compressed` type which in our case is the `G1Compressed` structure defined through the usage of the `new_curve_impl` macro in the `halo2` crate.

> As far as I know, all the necessary information about y coordinate is already stored in x.

So from what I get of your sentence and a few research, the compressed point is the data of `x` along with a sign bit that is derived from the `y` value of an affine point. And it seems that this is what we absorb in the Rust implementation. Aren't we missing the sign bit then? Or do we not need it?

Comment (storojs72) [reply src/blocks/IpaPcs.sol:294]: I think sign is also encoded in the information behind `x` (consider [from_bytes](https://github.com/privacy-scaling-explorations/halo2curves/blob/main/src/derive/curve.rs#L106) function in halo2curves). The field element for Bn256 (and Grumpkin) have 254 bits in size ([link](https://hackmd.io/@aztec-network/ByzgNxBfd#2-Grumpkin---A-curve-on-top-of-BN-254-for-SNARK-efficient-group-operations)), while `x` is `uint256`, which is 256 bits - so we have two bits for the metadata and by absorbing whole `uint256` we absorb sign as well. 

For the uncompressed commitment (which is essentially a Grumpkin affine point with two coordinates) we have different [absorb](https://github.com/lurk-lab/solidity-verifier/blob/main/src/blocks/KeccakTranscript.sol#L1235) where we put `x`, `y` and an extra-byte representing whether point is infinity (usually this is not the case, so this is 0x00) to the transcript's memory.

Comment (huitseeker) [reply src/blocks/IpaPcs.sol:294]: I believe the authoritative format for this conversion is in [SECG1](https://www.secg.org/sec1-v2.pdf), section 2.3.3. read with compression active.

**Pull request merged**